### PR TITLE
[SPARK-24470][Core] RestSubmissionClient to be robust against 404 & non json responses

### PR DESCRIPTION
**Pull request by rekhajoshm**
## What changes were proposed in this pull request?
Added check for 404, to avoid json parsing on not found response and to avoid returning malformed or bad request when it was a not found http response. 
Not sure if I need to add an additional check on non json response [if(connection.getHeaderField("Content-Type").contains("text/html")) then exception] as non-json is a subset of malformed json and covered in flow.

## How was this patch tested?
./dev/run-tests

### CODE REVIEW

**Comment by srowen** [comment]
Hm, I actually doubt it's ever correct to parse the result of `getErrorStream` here as JSON. It's going to be text or HTML document. The comment shows the intent was to return the error message to the user -- presumably as an exception as well, handled in the last line of the method. But it won't get there, right? JSON parsing fails before.

Therefore, what about throwing an exception right away on any non-200 response? Sure, try to put the content of `getErrorStream` in the message. `HistoryServerSuite` seems to do something like this.

**Review by srowen — APPROVED** [review]

**Comment by rekhajoshm** [comment]
```
Discovery starting.
Discovery completed in 5 seconds, 946 milliseconds.
Run starting. Expected test count is: 16
StandaloneRestSubmitSuite:
- construct submit request
- create submission
- create submission from main method
- kill submission
- request submission status
- create then kill
- create then request status
- create then kill then request status
- kill or request status before create
- good request paths
- good request paths, bad requests
- bad request paths
- server returns unknown fields
- client handles faulty server
- client does not send 'SPARK_ENV_LOADED' env var by default
- client includes mesos env vars
Run completed in 8 seconds, 480 milliseconds.
Total number of tests run: 16
Suites: completed 2, aborted 0
Tests: succeeded 16, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

File: core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala [hunk 1]
```diff
@@ -233,30 +233,41 @@ private[spark] class RestSubmissionClient(master: String) extends Logging {
...
+        logError(s"Server responded with error:\n${errString}")
+        val error = new ErrorResponse
+        if (responseCode == RestSubmissionServer.SC_UNKNOWN_PROTOCOL_VERSION) {
+          error.highestProtocolVersion = RestSubmissionServer.PROTOCOL_VERSION
         }
-      // If the server threw an exception while writing a response, it will not have a body
-      if (dataStream == null) {
-        throw new SubmitRestProtocolException("Server returned empty body")
+        error.message = errString.get
+        error
```
[the remark this answers is not in the document]
Comment (rekhajoshm) [reply core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala]: Yes, glad you see. i agree this is a clean design, serving our objective of robustly treating non-200 response, not sending them to json parsing and getting unneeded exceptions there.

File: core/src/test/scala/org/apache/spark/deploy/rest/StandaloneRestSubmitSuite.scala [hunk 2]
```diff
@@ -366,7 +366,8 @@ class StandaloneRestSubmitSuite extends SparkFunSuite with BeforeAndAfterEach {
     // server explodes internally beyond recovery
     // client should throw an appropriate exception to indicate server failure
     val conn3 = sendHttpRequest(statusRequestPath, "GET")
-    intercept[SubmitRestProtocolException] { client.readResponse(conn3) } // empty response
+    val response3 = client.readResponse(conn3)
```
[the remark this answers is not in the document]
Comment (rekhajoshm) [reply core/src/test/scala/org/apache/spark/deploy/rest/StandaloneRestSubmitSuite.scala]: Good point.Fixed.

File: core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala [hunk 3]
```diff
@@ -233,30 +233,43 @@ private[spark] class RestSubmissionClient(master: String) extends Logging {
...
-          connection.getInputStream
-        } else {
-          connection.getErrorStream
+      val responseCode = connection.getResponseCode
+
+      if (responseCode != HttpServletResponse.SC_OK) {
+        val errString = Some(Source.fromInputStream(connection.getErrorStream())
+          .getLines().mkString("\n"))
+        logError(s"Server responded with error:\n${errString}")
+        if (responseCode == HttpServletResponse.SC_INTERNAL_SERVER_ERROR && connection.getRequestMethod == "GET") {
```
[the remark this answers is not in the document]
Comment (rekhajoshm) [reply core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala]: Thanks @srowen  for the reviews. Fixed. As our intention is to also comply with existing tests, updated to verify with content_type.

The reason is we have two tests in "client handles faulty server" both of which return SC_INTERNAL_SERVER_ERROR but one is expected by tests to return ErrorMessage while other needs to be an exception.I found that strange. Anyhow, debugged and found one(KillSubmissionResponse) returns a valid json and other(StatusSubmissionResponse) is an html response.This is due to the internal behavior of KillRequestServlet and StatusRequestServlet.

This PR keeps tests behavior consistent, while addressing the objective of SPARK-24470

```
KillSubmissionResponse for KillSubmissionResponse is json/500 
{
  "action" : "ErrorResponse",
  "message" : "Internal server error: org.apache.spark.deploy.rest.SubmitRestProtocolException: Validation of message KillSubmissionResponse failed!\n\torg.apache.spark.deploy.rest.SubmitRestProtocolMessage.validate(SubmitRestProtocolMessage.scala:70)\n\torg.apache.spark.deploy.rest.RestServlet.validateResponse(RestSubmissionServer.scala:196)\n\torg.apache.spark.deploy.rest.RestServlet.sendResponse(RestSubmissionServer.scala:134)\n\torg.apache.spark.deploy.rest.KillRequestServlet.doPost(RestSubmissionServer.scala:223)\n\tjavax.servlet.http.HttpServlet.service(HttpServlet.java:707)\n\tjavax.servlet.http.HttpServlet.service(HttpServlet.java:790)\n\torg.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:848)\n\torg.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:584)\n\torg.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180)\n\torg.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)\n\torg.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)\n\torg.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)\n\torg.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)\n\torg.eclipse.jetty.server.Server.handle(Server.java:534)\n\torg.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:320)\n\torg.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)\n\torg.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:283)\n\torg.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:108)\n\torg.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93)\n\torg.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303)\n\torg.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148)\n\torg.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136)\n\torg.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671)\n\torg.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589)\n\tjava.lang.Thread.run(Thread.java:748)",
  "serverSparkVersion" : "2.4.0-SNAPSHOT"
}
```

```
StatusSubmissionResponse from /submissions/status/anything is html/500
<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 500 </title>
</head>
<body>
<h2>HTTP ERROR: 500</h2>
<p>Problem accessing /v1/submissions/status/anything. Reason:
<pre>    java.lang.ArithmeticException: / by zero</pre></p>
<hr /><a href="http://eclipse.org/jetty">Powered by Jetty:// 9.3.20.v20170531</a><hr/>
</body>
</html>
```

**Comment by rekhajoshm** [comment]
```
Discovery starting.
Discovery completed in 3 seconds, 75 milliseconds.
Run starting. Expected test count is: 16
StandaloneRestSubmitSuite:
- construct submit request
- create submission
- create submission from main method
- kill submission
- request submission status
- create then kill
- create then request status
- create then kill then request status
- kill or request status before create
- good request paths
- good request paths, bad requests
- bad request paths
- server returns unknown fields
- client handles faulty server
- client does not send 'SPARK_ENV_LOADED' env var by default
- client includes mesos env vars
Run completed in 5 seconds, 483 milliseconds.
Total number of tests run: 16
Suites: completed 2, aborted 0
Tests: succeeded 16, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

**Review by srowen — APPROVED** [review]
Looks like this preserves behavior and improves the error message.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**